okhttp/src/main/java/okhttp3/internal/http2/Http2Writer.kt

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright (C) 2011 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.http2
+
+import okhttp3.internal.Util.format
+import okhttp3.internal.http2.Http2.CONNECTION_PREFACE
+import okhttp3.internal.http2.Http2.FLAG_ACK
+import okhttp3.internal.http2.Http2.FLAG_END_HEADERS
+import okhttp3.internal.http2.Http2.FLAG_END_STREAM
+import okhttp3.internal.http2.Http2.FLAG_NONE
+import okhttp3.internal.http2.Http2.INITIAL_MAX_FRAME_SIZE
+import okhttp3.internal.http2.Http2.TYPE_CONTINUATION
+import okhttp3.internal.http2.Http2.TYPE_DATA
+import okhttp3.internal.http2.Http2.TYPE_GOAWAY
+import okhttp3.internal.http2.Http2.TYPE_HEADERS
+import okhttp3.internal.http2.Http2.TYPE_PING
+import okhttp3.internal.http2.Http2.TYPE_PUSH_PROMISE
+import okhttp3.internal.http2.Http2.TYPE_RST_STREAM
+import okhttp3.internal.http2.Http2.TYPE_SETTINGS
+import okhttp3.internal.http2.Http2.TYPE_WINDOW_UPDATE
+import okhttp3.internal.http2.Http2.frameLog
+import okhttp3.internal.writeMedium
+import okio.Buffer
+import okio.BufferedSink
+import java.io.Closeable
+import java.io.IOException
+import java.util.logging.Level.FINE
+import java.util.logging.Logger
+
+/** Writes HTTP/2 transport frames.  */
+internal class Http2Writer(
+  private val sink: BufferedSink,
+  private val client: Boolean
+) : Closeable {
+  private val hpackBuffer: Buffer = Buffer()
+  private var maxFrameSize: Int = INITIAL_MAX_FRAME_SIZE
+  private var closed: Boolean = false
+  val hpackWriter: Hpack.Writer = Hpack.Writer(hpackBuffer)
+
+  @Synchronized @Throws(IOException::class)
+  fun connectionPreface() {
+    if (closed) throw IOException("closed")
+    if (!client) return // Nothing to write; servers don't send connection headers!
+    if (logger.isLoggable(FINE)) {
+      logger.fine(format(">> CONNECTION %s", CONNECTION_PREFACE.hex()))
+    }
+    sink.write(CONNECTION_PREFACE.toByteArray())
```

Comment:
unrelated to conversion: `toByteArray()` shouldn't be needed? I can do a pass of this over the entire codebase once merged.

Reply:
oooh, nice. Fixed.